core/block/backlinks/watcher.go

```diff
@@ -67,109 +75,169 @@ func (uw *UpdateWatcher) Close(context.Context) error {
+
+			applyUpdates(accumulatedBacklinks, info)
 		}
+		lastReceivedUpdates = time.Now()
+		l.Unlock()
 	}
 }
 
-func (uw *UpdateWatcher) updateBackLinksInObjects(info objectstore.BacklinksUpdateInfo) {
-	spaceId, err := uw.resolver.ResolveSpaceID(info.Id)
+func (uw *UpdateWatcher) updateBackLinksInObject(id string, backlinksUpdate *backLinksUpdate) {
+	spaceId, err := uw.resolver.ResolveSpaceID(id)
 	if err != nil {
-		log.With("objectID", info.Id).Errorf("failed to resolve space id for object %s: %v", info.Id, err)
+		log.With("objectID", id).Errorf("failed to resolve space id for object %s: %v", id, err)
 		return
 	}
 	spc, err := uw.spaceService.Get(context.Background(), spaceId)
 	if err != nil {
-		log.With("objectID", info.Id).Errorf("failed get space %s: %v", spaceId, err)
+		log.With("objectID", id).Errorf("failed get space %s: %v", spaceId, err)
 		return
 	}
 
-	addBacklink := func(current *types.Struct) (*types.Struct, error) {
+	updateBacklinks := func(current *types.Struct, backlinksChange *backLinksUpdate) (*types.Struct, error) {
 		if current == nil || current.Fields == nil {
 			return nil, objectstore.ErrDetailsNotChanged
 		}
 		backlinks := pbtypes.GetStringList(current, bundle.RelationKeyBacklinks.String())
-		if lo.Contains(backlinks, info.Id) {
-			return nil, objectstore.ErrDetailsNotChanged
-		}
-		backlinks = append(backlinks, info.Id)
-		current.Fields[bundle.RelationKeyBacklinks.String()] = pbtypes.StringList(backlinks)
-		return current, nil
-	}
 
-	removeBacklink := func(current *types.Struct) (*types.Struct, error) {
-		if current == nil || current.Fields == nil {
-			return nil, objectstore.ErrDetailsNotChanged
+		for _, removed := range backlinksChange.removed {
+			slice.Remove(backlinks, removed)
 		}
-		backlinks := pbtypes.GetStringList(current, bundle.RelationKeyBacklinks.String())
-		newBacklinks := slice.Remove(backlinks, info.Id)
-		if len(backlinks) == len(newBacklinks) {
-			return nil, objectstore.ErrDetailsNotChanged
+
+		for _, added := range backlinksChange.added {
+			if !lo.Contains(backlinks, added) {
+				backlinks = append(backlinks, added)
+			}
 		}
-		current.Fields[bundle.RelationKeyBacklinks.String()] = pbtypes.StringList(newBacklinks)
+
+		current.Fields[bundle.RelationKeyBacklinks.String()] = pbtypes.StringList(backlinks)
 		return current, nil
 	}
 
-	for _, modification := range []struct {
-		ids      []string
-		modifier func(details *types.Struct) (*types.Struct, error)
-	}{
-		{info.Added, addBacklink},
-		{info.Removed, removeBacklink},
-	} {
-		for _, id := range modification.ids {
-			err = spc.DoLockedIfNotExists(id, func() error {
-				return uw.store.ModifyObjectDetails(id, modification.modifier)
+	err = spc.DoLockedIfNotExists(id, func() error {
+		return uw.store.ModifyObjectDetails(id, func(details *types.Struct) (*types.Struct, error) {
+			return updateBacklinks(details, backlinksUpdate)
+		})
+	})
+
+	if !errors.Is(err, ocache.ErrExists) {
```

Comment:
If there is no error shouldn't you return from func, i.e. if you applied the changes when the object was not in cache?

Reply:
BTW yesterday I found many cases when error is DetailsNotChanged. That is absolutely harmless error, so should be considered as normal